### PR TITLE
Add more messages to a VM lifecycle

### PR DESCRIPTION
**Pull request by benjamreis**
Signed-off-by: BenjiReis <benjamin.reis@vates.fr>

### CODE REVIEW

**Comment by benjamreis** [comment]
Solves https://github.com/xapi-project/xen-api/issues/4393

File: ocaml/xapi/message_forwarding.ml [hunk 1]
```diff
@@ -2160,7 +2210,20 @@ functor
...
+            (Db.VM.get_name_label ~__context ~self:snapshot)
+        in
+        let name, priority = Api_messages.vm_snapshot_reverted in
+        try
+          ignore
+            (Xapi_message.create ~__context ~name ~priority ~cls:`VM
+               ~obj_uuid:(Db.VM.get_uuid ~__context ~self:vm)
+               ~body:message_body
+            )
+        with _ -> ()
```
Comment (psafont) [inline ocaml/xapi/message_forwarding.ml:2226]: There is quite a bit of repetition here, it feels like creating a function that is used in all these sites is worth it, something like:
```ocaml
let create_vm_message create_body message vm =
  let message_body = create_body () in
  let name, priority = Api_messages.vm_snapshot_reverted in
  try
    ignore
      (Xapi_message.create ~__context ~name ~priority ~cls:`VM
         ~obj_uuid:(Db.VM.get_uuid ~__context ~self:vm)
         ~body:message_body
      )
    with _ -> ()

...

let create_body () = 
  Printf.sprintf "VM '%s' reverted to snapshot: %s"
    (Db.VM.get_name_label ~__context ~self:vm)
    (Db.VM.get_name_label ~__context ~self:snapshot)
in
create_vm_message create_body Api_messages.vm_snapshot_reverted vm
```

File: ocaml/xapi/message_forwarding.ml [hunk 2]
```diff
@@ -1053,6 +1053,16 @@ functor
 
       let wait_for_tasks = Helpers.Task.wait_for
 
+      let create_vm_message ~__context ~body ~message ~vm =
```
Comment (psafont) [inline ocaml/xapi/message_forwarding.ml:1056]:
```suggestion
      let create_vm_message ~__context ~vm ~message_body ~message =
```
Changing the order and the name of the content parameter will make call sites look more straight-forward:

```ocaml
        create_vm_message ~__context ~vm ~message_body
          ~message:Api_messages.vm_checkpointed;
```

**Review by psafont — APPROVED** [review]
Small nit aside, this looks good to me. This can help users to easily see VM operation being completed.

File: ocaml/xapi/message_forwarding.ml [hunk 3]
```diff
@@ -2328,6 +2302,14 @@ functor
                 )
             )
         ) ;
+        let message_body =
+          Printf.sprintf "VM '%s' migrated from host '%s' to host '%s'"
+            (Db.VM.get_name_label ~__context ~self:vm)
+            (Db.Host.get_name_label ~__context ~self:source_host)
```
Comment (lindig) [inline ocaml/xapi/message_forwarding.ml:2308]: Should this report the UUIDs, too?

**Review by lindig — APPROVED** [review]

File: ocaml/xapi/message_forwarding.ml (hunk 3 above)
Comment (benjamreis) [reply ocaml/xapi/message_forwarding.ml:2308]: I can add it, the uuid was not in pre-existing messages, but IMHO adding it is a good idea as we have no insurance name_label is unique per vm. Should i do it?

Comment (psafont) [reply ocaml/xapi/message_forwarding.ml:2308]: Yeah, go for it, I'm sure @stormi will like that

Comment (benjamreis) [reply ocaml/xapi/message_forwarding.ml:2328]: Done :+1: I've added the uuids of the hosts in the migration messages as well :)

**Comment by psafont** [comment]
Could you squash the commits and add to the commit message why the messages are being added?

File: ocaml/xapi-consts/api_messages.ml [hunk 4]
```diff
@@ -84,12 +84,24 @@ let vm_shutdown = addMessage "VM_SHUTDOWN" 5L
 
 let vm_rebooted = addMessage "VM_REBOOTED" 5L
 
+let vm_migrated = addMessage "VM_MIGRATED" 5L
+
+let vm_snapshoted = addMessage "VM_SNAPSHOTED" 5L
```
Comment (robhoes) [inline ocaml/xapi-consts/api_messages.ml:89]: This looks a little odd... I think it should be "snapshotted".

Comment (benjamreis) [reply ocaml/xapi-consts/api_messages.ml]: Indeed, it's fixed!

**Review by robhoes — APPROVED** [review]
I think that this is a useful addition.

**Comment by edwintorok** [comment]
This has caused SXM failures, see https://github.com/xapi-project/xen-api/pull/4448

**Pull request merged**